internal/distro/distro.go

```diff
@@ -275,3 +275,49 @@ func ExportsFallback() []string {
 func PayloadPackageSets() []string {
 	return []string{}
 }
+
+func MakePackageSetChains(t ImageType, packageSets map[string]rpmmd.PackageSet, repos []rpmmd.RepoConfig) map[string][]rpmmd.PackageSet {
+	allSetNames := make([]string, len(packageSets))
+	idx := 0
+	for setName := range packageSets {
+		allSetNames[idx] = setName
+		idx++
+	}
+
+	// map repository PackageSets to the list of repo configs
+	packageSetsRepos := make(map[string][]rpmmd.RepoConfig)
+	for idx := range repos {
+		repo := repos[idx]
+		if len(repo.PackageSets) == 0 {
+			// repos that don't specify package sets get used everywhere
+			repo.PackageSets = allSetNames
+		}
+		for _, name := range repo.PackageSets {
+			psRepos := packageSetsRepos[name]
+			psRepos = append(psRepos, repo)
+			packageSetsRepos[name] = psRepos
+		}
+	}
+
+	chainedSets := make(map[string][]rpmmd.PackageSet)
+	// first collect package sets that are part of a chain
+	for specName, setNames := range t.PackageSetsChains() {
+		pkgSets := make([]rpmmd.PackageSet, len(setNames))
+
+		// add package-set-specific repositories to each set if one is defined
+		for idx, pkgSetName := range setNames {
+			pkgSets[idx] = packageSets[pkgSetName]
+			pkgSets[idx].Repositories = packageSetsRepos[pkgSetName]
+			delete(packageSets, pkgSetName)
```

Comment:
> The fix should be simply to not delete the package set from `packageSets`, rather in the `for` cycle at the end of this function to check if the package set name is already contained in `chainedSets` and if yes, then `continue`.

Doing the code now this feels wrong.  We're already iterating all the package set names in the chains so it feels weird to then have to scan through them again for every key in the package sets at the end.  I know we usually only have one chain with two package sets in it, but I still don't like doing unnecessary iterations through lists when it can be avoided.

I do think that it's wrong to delete the package set from the map, because the map is the argument and we don't know if the caller will want to reuse it.  I'm going to make a set of "seen" package sets instead.

Reply:
I didn't mean it literally to iterate over all the package set names again, but actually to use map in exactly the way you used it. Not surprisingly, your end solution is exactly the same as I hoped for it to be 😄